README.md
# Emotiv SDK - Community Edition

Here is the repository of the Emotiv SDK Community Edition, along with code examples and other development tools.
* Supports EPOC (over Emotiv USB dongle)
* Supports Insight and EPOC+ (over BTLE and Emotiv USB dongle)
* Supports Win32/64, Mac, iOS, Android, iOS, Ubuntu and Debian (armhf)

If you have questions or have knowledge to share, please visit our [forum](https://emotiv.com/forum/) which is the hub for our developer community.

To get the SDK you can either:
* Clone this repository with `git` command line tool or a GUI client like [SourceTree](https://www.sourcetreeapp.com); or
* Click on the "Download ZIP" button (above, on the right) to get the zip archive of the latest version; or
* Go to [releases](https://github.com/Emotiv/community-sdk/releases) and download the archive of a particular release

This repo contains SDK for Win32/64, Mac, iOS, Ubuntu and Debian (armhf).

Android SDK will be updated on [jcenter](https://bintray.com/nguoithichkhampha/maven/community) separately.

## Table of Contents
1. [Changelog](#changelog)
2. [Supported Platforms](#supported-platforms)
3. [Connecting with BTLE](#connecting-with-btle)
4. [API Documentation](#api-documentation)
5. [API Code Example](#api-code-example)
6. [Apps and Tools](#apps-and-tools)
7. [Support](#support)

## Changelog
See [changelog](CHANGELOG.md)

## Supported Platforms
* Insight with Bluetooth SMART (Bluetooth 4.0 or Bluetooth Low Energy) connection currently works with the SDK under:
  * Windows 8 or above
  * Windows 7, **only with Bluetooth adaptor that uses Broadcom BCM20702 chipset**
    (recommended one: Plugable USB 2.0 Bluetooth Adapter, http://plugable.com/products/usb-bt4le)
  * Mac OS X 10.10 or above (check if Bluetooth LMP Version is 0x6 from System Report)
  * iOS 8.0 or above (excluding iOS 8.3 and 8.4)
  * Android 4.4.3 or above (excluding Android 5.0)
  * Ubuntu 16.04 (see [Connecting on Ubuntu and Debian](doc/Connecting%20on%20Ubuntu%20and%20Debian.md))
  * Debian 8.3 Jessie (armhf) (see [Connecting on Ubuntu and Debian](doc/Connecting%20on%20Ubuntu%20and%20Debian.md))

* Insight with Emotiv Universal USB Receiver currently works with the SDK under:
  * Windows 7 or above
  * Mac OS X 10.8 or above
  * Android 4.4.3 or above
  * Ubuntu 16.04
  * Debian 8.3 Jessie (armhf)

## Connecting with BTLE
* Windows: Turn on Bluetooth on both Insight and PC, then pair your Insight with Windows built-in Bluetooth service first
* Mac, Android and iOS: Turn on Bluetooth on both Insight and Mac/Android device, then start Emotiv app to use (without first pairing)
* Ubuntu and Debian: See [Connecting on Ubuntu and Debian](doc/Connecting%20on%20Ubuntu%20and%20Debian.md)

## API Documentation
http://emotiv.github.io/community-sdk/

## API Code Example

There are 2 groups of examples :
 * Examples which could run without SDK license are placed inside folder **examples_basic**
 * Examples which need to have SDK license to run are placed inside folder **examples_extra_prime**.
   In case you want to use these functionality in your development, please [contact us for license](https://www.emotiv.com/sdk-premium-edition-application-form/)

Table below describes examples in detail :

| Name  | Could run without license |Available with language|Support Platform|
|---|---|---|---|
| AverageBandPowers (FFTSample) | Yes | C++, C#, Java, Python, Objective-C, Matlab |Win, Mac, Ubuntu, Android, iOS|
| EmoStateLogger | Yes | C++, C#, Java, Python | Win, Mac, Ubuntu |
| FacialExpressionDemo | Yes | C++, Matlab | Win, Mac, Ubuntu |
| FacialExpression | Yes | Objective-C, Java, Matlab | Mac, iOS, Android |
| GyroData | Yes | C++ | Win, Mac, Ubuntu |
| HeadsetInformationLogger | Yes | C++, C#, Java, Python | Win, Mac, Ubuntu |
| MentalCommand | Yes | C++, Objective-C, Java | Win, Mac, Ubuntu, Android, iOS |
| MentalCommandWithCloudProfile | Yes | C++, C#, Java, Python | Win, Mac, Ubuntu |
| MentalCommandWithLocalProfile | Yes | C++, C# | Win, Mac |
| MotionDataLogger | Yes | C++, C#, Java, Python, Objective-C, Matlab | Win, Mac, Ubuntu, Android, iOS |
| MultiDongleConnection | Yes | C++, C#, Python | Win, Mac, Ubuntu |
| SavingAndLoadingProfileCloud | Yes | C++, C#, Java, Python |Win, Mac, Ubuntu|
| ProfileCloudExample | Yes | Objective-C, Java | Mac, Android, iOS|
| RollingBall  | Yes | Unity | Win, Mac |
| ActivateLicense  | No | C++, C#, Objective-C, Java, Python, Matlab | Win, Mac, Android |
| EEGLogger  | No | C++, C#, Objective-C, Java, Python, Matlab | Win, Mac, Android |
| MultiDongleEEGLogger  | No | C++ | Win, Mac |
| MultilChannelEEGLogger  | No | C++ | Win, Mac |
| PerformanceMetricsDemo  | Yes | C++, C#, Python, Matlab | Win, Mac |
| EmotionalState  | No | Objective-C | iOS, Mac |

Doesn't support Python 64bit. 

## Apps and Tools

#### Xavier EmoComposer
 * [Xavier Composer](tools) allows you to send user-defined EmoStates to Emotiv SDK, Xavier EmoKey, or any other application that makes use of the Emotiv API, and simulate the events that could be sent from Emotiv SDK with a real device.
 * Available for Windows 7 or above and Mac OS X 10.8 or above

#### Xavier EmoKey
 * [Xavier EmoKey](tools) translates Emotiv detection results to predefined sequences of keystrokes according to logical rules defined by the user.
 * Available for Windows 7 or above and Mac OS X 10.8 or above

#### MyEmotiv App for Android
Requires Android 4.4.3 or later

https://play.google.com/store/apps/details?id=com.emotiv.insightapp

#### MyEmotiv App for iOS
Requires iOS 8.0 or later, **excluding iOS 8.3/8.4 due to BTLE data transmission limitation in those versions**

https://itunes.apple.com/us/app/emotiv-insight/id1031805596

## Support

Please check out the topic **Insight** on **Emotiv Help Centre**:

https://emotiv.zendesk.com/hc/en-us/categories/200100495-Insight

Our knowledge base is a good source for further reading:

https://emotiv.zendesk.com/hc/en-us
 
Please also visit our [forum](https://emotiv.com/forum/) for bug reports and feature requests.

Happy coding!

The Emotiv Team


CHANGELOG.md
## Emotiv SDK Community Edition Changelog

SDK CHANGELOGS

#### Version 3.5.1
* New features and improvement:
    * Support Android for SDK 3.5.1 and introduce some android examples
    * Update new license model for sdk 3.5.1. Removed daily and monthly debit limit.
	  support annual license
    * Update EmotivTools for version 3.5.1 both Windows and MAC        
    * Update emostatelogger examples for Csharp and Java
    * Introduce EdfHeaderConverter tool for mac
    
* Bug fixes:
    * Fixed error "Expried license" after renew license key
    * Fixed issue of MentalCommandWithCloudProfile - python example
	
* Platforms:
    * SDK v3.5.1 support Windows, MAC OS, Android
    * Linux, IOS: still use SDK version 3.3.3

#### Version 3.5.0
* New features and improvement:
    * Renamed error code `EDK_PROFILE_CLOUD_EXISTED` to `EDK_CLOUD_PROFILE_EXISTS`
    * Removed `EDK_OVER_QUOTA_IN_DAY` and `EDK_OVER_QUOTA_IN_MONTH` error codes
    * Improved samples interpolation during packet lost
    * Improved noise handling in Performance Metric detections with Insight headset
    * Updated results of `IEE_GetAverageBandPowers()` from 2Hz to 8Hz
    * Improved accuracy of timestamp in EEG / EmoState data streams
    * New API: `IEE_DataEnableBufferForMarker()` which enables EEG/Motion buffer and allows user to set marker with a specific timestamp
    * Fixed EdfHeaderConverter tool to work with new EDF version (20170703)
      https://github.com/Emotiv/community-sdk/tree/master/tools/EEGLAB_EDF_FIXED
    * Data stream type now included in `userID` as a bit mask:
    ```
      DONGLE_STREAM_MASK   = 0x1000;
      BTLE_STREAM_MASK     = 0x2000;
      EDF_STREAM_MASK      = 0x3000;
      CSV_STREAM_MASK      = 0x4000;
      MEM_STREAM_MASK      = 0x5000;       
      EXTENDER_STREAM_MASK = 0x6000;
     ```
    * Support new licensing model - for details https://www.emotiv.com/developer/
* Platforms: 
    * SDK v3.5.0 support Windows, MAC OS
    * Linux, Android, IOS: still use SDK version 3.3.3
    
#### Version 3.4.0
* Bug fixes:
    * Fixed bug: cannot login with special symbol in EmotivID or password
    * Fixed bug: battery level is not stable	
	
* New features and improvement:
    * Support legacy EDK, which includes `EE_` API
    * New License API in `EmotivLicense.h`- allow accessing EEG and Performance Metric with license key
    * New Local Training Profile API in `IEmotivProfile.h` - allow saving/loading a training profile to local device
    * New Cloud Training Profile API in `EmotivCloudClient.h`
    * Improved timestamp accuracy for EEG and Motion data
    * Improved marker timestamp accuracy
    * Changed `EC_UpdateUserProfile` to return `EDK_` error code
    * New API: `IEE_HeadsetGetSerialNumber` and `IEE_DisconnectDevice()`
    * Merged `EC_` error code in `EmotivCloudErrorCode.h` into `EDK_` error code in `IedkErrorCode.h`
    * Deprecated `EmotivCloudErrorCode.h`
    * Removed `IEE_GetInsightSignalStrength()` and `IEE_GetEpocPlusSignalStrength()`
    * Added `PM_INTEREST` and `PM_FOCUS` for `IS_PerformanceMetricIsActive()`
	
* Known issues:
    * BTLE connection is not stable on Windows 10 due to intermediate packet lost
    
* Platforms: 
    * SDK v3.4.0 support Windows, MAC OS
    * Linux, Android, IOS: still use SDK version 3.3.3

#### Version 3.3.3
* Bug fixes:
  * Fixed bug in `IS_GetHeadsetOn` and `IS_GetBatteryChargeLevel` which always return 1
  * Fixed bug that prevents connecting headset with firmware version `0x5xx`
  * Fixed `IEE_EngineRemoteConnect()` connection from iOS to Mac (pending on other platforms)
* New features and improvement:
  * New API: `IEE_HeadsetGetSerialNumber`
  * Speed optimization with EmotivCloudClient API and it now returns error code for easier debugging
  * Support iOS Simulator (x86_64 and i386 architectures)

#### Version 3.3.2
* Bug fixes:
  * Fixed unexpected failure in Mental Commands training
  * Fixed connection error for examples on Mac
* Improvement:
  * Support cmake on Ubuntu and Debian/armhf in examples
  * Add BlueAvatar and EmoCube for examples on Mac

#### Version 3.3.1
* New features and improvement:
  * Improved BTLE connection on all platforms
  * Improved noise tolerance for Mental Commands with Insight
  * Added `IEE_GetHeadsetSettings` and `IEE_SetHeadsetSettings` for EPOC+
  * Added `EC_ReconnectEngine` and `EC_DisconnectEngine` in EmotivCloudClient.h
  * Added `FE_LAUGH`, `FE_SMIRK_LEFT`, `FE_SMIRK_RIGHT` into detection list of Facial Expression
  * Added `IS_FacialExpressionIsLookingLeft`, `IS_FacialExpressionIsLookingRight` for EPOC/EPOC+.
  * Removed param `profileName` from `EC_UpdateUserProfile`
  * Added `IEE_GetInsightDeviceState` and `IEE_GetEpocPlusDeviceState` (Mac, iOS, Android)
  * Added multi-app support (Mac, iOS, Android)
  * Supports Ubuntu 15.10 64-bit
  * Supports Debian 8.3 Jessie (armhf) on Raspberry Pi 2
* Bug fixes:
  * Fixed bug when uploading/downloading profile on 64-bit platform
  * Fixed incorrect battery level display with Insight (require new firmware )
  * Fixed bug with Surprise and Frown detections
  * Fixed motion data values retrieved from EPOC+
  * Fixed bug that prevent Insight to work with Extender on Mac
  * Fixed bug that prevent connecting to headset via BTLE on Linux
  * Swapped Gyro Y and Z in motion data with Insight
* Known Issues:
  * EPOC+ motion data may have packet lost on Android
  * EPOC+ only works with dongle version `0x06ff`, not with `0x1fff` or `0x0565`
  * Changing the settings on EPOC+ with firmware `0x610` may fail and eventually turns the headset into a brick (!). Please contact hello@emotiv.com to arrange for a firmware upgrade

#### Version 3.3.0
* Supports saving/loading profile to Emotiv Cloud (see EmotivCloudClient.h)
* New API `IEE_GetAverageBandPowers` to get average band powers for:
  * theta (4-8 Hz)
  * alpha (8-12 Hz)
  * low-beta (12-16 Hz)
  * high-beta (16-25 Hz)
  * gamma (25-45 Hz)
* Supports Ubuntu 14.04 64-bit

#### Version 3.2.0

examples_extra_prime/Matlab/eeglogger.m
%  ** Copyright 2017 by Emotiv. All rights reserved
%  ** Example - EEG Logger
% Require Emotiv Advanced/Prime SDK or EmotivPRO license activated.
% Data is read from the headset and sent to an output file for later analysis
% Check result file: eegloger.csv
clc;
w = warning ('off','all');
bitVersion = computer('arch');
if (strcmp(bitVersion, 'win64'))
    loadlibrary('../../bin/win64/edk.dll','../../include/Iedk.h','addheader','IedkErrorCode.h','addheader','IEmoStateDLL.h','addheader','FacialExpressionDetection.h','addheader','MentalCommandDetection.h','addheader','IEmotivProfile.h','addheader','EmotivLicense.h','alias','libIEDK');
    loadlibrary('../../bin/win64/edk.dll','../../include/IEegData.h','addheader','Iedk.h','alias','libIEEGDATA');
else
    loadlibrary('../../bin/win32/edk.dll','../../include/Iedk.h','addheader','IedkErrorCode.h','addheader','IEmoStateDLL.h','addheader','FacialExpressionDetection.h','addheader','MentalCommandDetection.h','addheader','IEmotivProfile.h','addheader','EmotivLicense.h','alias','libIEDK');
    loadlibrary('../../bin/win32/edk.dll','../../include/IEegData.h','addheader','Iedk.h','alias','libIEEGDATA');
end

enuminfo.IEE_DataChannels_enum = struct('IED_COUNTER', 0, 'IED_INTERPOLATED', 1, 'IED_RAW_CQ', 2,'IED_AF3', 3, 'IED_F7',4, 'IED_F3', 5, 'IED_FC5', 6, 'IED_T7', 7,'IED_P7', 8, 'IED_Pz', 9,'IED_O2', 10, 'IED_P8', 11, 'IED_T8', 12, 'IED_FC6', 13, 'IED_F4', 14, 'IED_F8', 15, 'IED_AF4', 16, 'IED_GYROX', 17,'IED_GYROY', 18, 'IED_TIMESTAMP', 19,'IED_MARKER_HARDWARE', 20, 'IED_ES_TIMESTAMP',21, 'IED_FUNC_ID', 22, 'IED_FUNC_VALUE', 23, 'IED_MARKER', 24,'IED_SYNC_SIGNAL', 25);
enuminfo.IEE_MentalCommandTrainingControl_enum = struct('MC_NONE',0,'MC_START',1,'MC_ACCEPT',2,'MC_REJECT',3,'MC_ERASE',4,'MC_RESET',5);

DataChannels = enuminfo.IEE_DataChannels_enum;
DataChannelsNames = {'IED_COUNTER','IED_INTERPOLATED','IED_AF3','IED_T7','IED_Pz','IED_T8','IED_AF4','IED_GYROX','IED_GYROY','IED_TIMESTAMP','IED_ES_TIMESTAMP'};
DataChannelsNamesfull ={'IED_COUNTER','IED_INTERPOLATED','IED_RAW_CQ','IED_AF3','IED_F7','IED_F3','IED_FC5','IED_T7','IED_P7','IED_Pz','IED_O2','IED_P8','IED_T8','IED_FC6','IED_F4','IED_F8','IED_AF4','IED_GYROX','IED_GYROY','IED_TIMESTAMP','IED_MARKER_HARDWARE','IED_ES_TIMESTAMP','IED_FUNC_ID','IED_FUNC_VALUE','IED_MARKER','IED_SYNC_SIGNAL'};


% libfunctionsview('libIEDK');
% libfunctionsview('libIEEGDATA');

fid = fopen('eegloger.csv','wt');
header = 'IED_COUNTER,IED_INTERPOLATED,IED_RAW_CQ,IED_AF3,IED_F7,IED_F3,IED_FC5,IED_T7,IED_P7,IED_Pz,IED_O2,IED_P8,IED_T8,IED_FC6,IED_F4,IED_F8,IED_AF4,IED_GYROX,IED_GYROY,IED_TIMESTAMP,IED_MARKER_HARDWARE,IED_ES_TIMESTAMP,IED_FUNC_ID,IED_FUNC_VALUE,IED_MARKER,IED_SYNC_SIGNAL\n';

fprintf(fid,header);
EDK_OK = 0;
calllib('libIEDK','IEE_EngineConnect', 'Emotiv Systems-5');

eEvent = calllib('libIEDK','IEE_EmoEngineEventCreate');
bufferSizeInSec = 0.05;
hData = calllib('libIEEGDATA','IEE_DataCreate');
calllib('libIEEGDATA','IEE_DataSetBufferSizeInSec', bufferSizeInSec)
readytocollect = false;
acqtime = 10;
% initialize outputs:

tic;
mycolumn = numel(DataChannelsNamesfull);
fprintf('Run time: %d seconds', acqtime);
while (toc<acqtime)
    
    state = calllib('libIEDK','IEE_EngineGetNextEvent', eEvent); % state = 0 if everything's OK
    
    if(state == EDK_OK)
        eventType = calllib('libIEDK','IEE_EmoEngineEventGetType',eEvent);
        userID=libpointer('uint32Ptr',0);
        calllib('libIEDK','IEE_EmoEngineEventGetUserId',eEvent, userID);
        
        if (strcmp(eventType,'IEE_UserAdded') == true)
            userID_value = get(userID,'value');
            calllib('libIEEGDATA','IEE_DataAcquisitionEnable',userID_value,true);
            readytocollect = true;
        end
    end
    
    if(readytocollect)
        result = calllib('libIEEGDATA','IEE_DataUpdateHandle', userID_value, hData);
        
        if(result~=0)
            continue;
        end
        nSamples = libpointer('uint32Ptr',0);
        calllib('libIEEGDATA','IEE_DataGetNumberOfSample', hData, nSamples);
        nSamplesTaken = get(nSamples,'value');
        
        if (nSamplesTaken ~= 0)
            data = libpointer('doublePtr', zeros(1, nSamplesTaken));
            data2=zeros(nSamplesTaken,numel(DataChannelsNames));
            
            for i = 1:mycolumn
                calllib('libIEEGDATA', 'IEE_DataGet', hData, DataChannels.([DataChannelsNamesfull{i}]), data, uint32(nSamplesTaken));
                data_value = get(data,'value');
                for k = 1: nSamplesTaken
                    data2(k,i) = data_value(k);
                end
                disp(data_value);
            end
            
            for i = 1 : nSamplesTaken
                dlmwrite('eegloger.csv', data2(i,1:mycolumn),'-append','delimiter',',','precision','%.2f');
            end
        end
        
    end
    
end

delete data2;
fclose(fid);
clearvars userID;
clearvars nSamples;
clearvars data;

calllib('libIEEGDATA','IEE_DataFree',hData);
calllib('libIEDK','IEE_EngineDisconnect')
calllib('libIEDK','IEE_EmoEngineEventFree', eEvent);
disp('finish');

